gtsam/hybrid/HybridGaussianConditional.cpp

```diff
@@ -28,56 +28,60 @@
 #include <gtsam/linear/GaussianFactorGraph.h>
 
 namespace gtsam {
-HybridGaussianFactor::FactorValuePairs GetFactorValuePairs(
-    const HybridGaussianConditional::Conditionals &conditionals) {
-  auto func = [](const GaussianConditional::shared_ptr &conditional)
-      -> GaussianFactorValuePair {
-    double value = 0.0;
-    // Check if conditional is pruned
-    if (conditional) {
-      // Assign log(\sqrt(|2πΣ|)) = -log(1 / sqrt(|2πΣ|))
-      value = conditional->negLogConstant();
-    }
-    return {std::dynamic_pointer_cast<GaussianFactor>(conditional), value};
-  };
-  return HybridGaussianFactor::FactorValuePairs(conditionals, func);
-}
+/* *******************************************************************************/
+struct HybridGaussianConditional::ConstructorHelper {
+  KeyVector frontals, parents;
+  HybridGaussianFactor::FactorValuePairs pairs;
+  double negLogConstant;
+  /// Compute all variables needed for the private constructor below.
+  ConstructorHelper(const Conditionals &conditionals) {
+    negLogConstant = std::numeric_limits<double>::infinity();
+
+    auto func = [&](const GaussianConditional::shared_ptr &c)
+        -> GaussianFactorValuePair {
+      double value = 0.0;
+      if (c) {
+        if (frontals.empty()) {
+          frontals = KeyVector(c->frontals().begin(), c->frontals().end());
+          parents = KeyVector(c->parents().begin(), c->parents().end());
+        }
+        value = c->negLogConstant();
+        negLogConstant = std::min(negLogConstant, value);
+      }
+      return {std::dynamic_pointer_cast<GaussianFactor>(c), value};
+    };
+    pairs = HybridGaussianFactor::FactorValuePairs(conditionals, func);
+  }
+};
+
+/* *******************************************************************************/
+HybridGaussianConditional::HybridGaussianConditional(
+    const DiscreteKeys &discreteParents,
+    const HybridGaussianConditional::Conditionals &conditionals,
+    const ConstructorHelper &helper)
+    : BaseFactor(discreteParents, helper.pairs),
+      BaseConditional(helper.frontals.size()),
+      conditionals_(conditionals),
```

Comment:
These things are here twice! Once in BaseFactor, and once in conditionals_.

Reply:
Yeah I thought about that too. There was a reason I couldn't get rid of `conditionals_`, which I think was to do with `negLogConstant`. I'll think about this again so we can remove it.